srk/src/quantifier.ml

```diff
@@ -2143,3 +2148,306 @@ let local_project_cube srk exists model cube =
+    | `Ite (cond, bif, belse) ->
+      let vars1, _, _ = cond in
+      let vars2, _, _ = bif in
+      let vars3, _, _ = belse in
+      DecMap.union (DecMap.union vars1 vars2) vars3, 0, `Ite(cond, bif, belse)
+  in
+  let mk_junc op =
+    match op with
+    | `Or -> mk_or srk
+    | `And -> mk_and srk
+  in
+  let mk_q op =
+    match op with
+    | `Exists -> mk_exists srk
+    | `Forall -> mk_forall srk
+  in
+  let q_map = BatHashtbl.create 97 in
+  (* This function the constructs a formula *)
+  let rec compute_expr tot_del depth node =
+    let vars,q_del, content = node in
+    let c = DecMap.get_dec vars in
+    let tot_del' = tot_del + q_del in
+    (* The idea is that we place quantifier info in a hashtbl at the quantifiers
+     * depth in the original expression (where the outer-most quantifiers
+     * have depth 0).*)
+    match content with
+    | `Tru -> mk_true srk
+    | `Fls -> mk_false srk
+    | `Atom phi | `Prop phi ->
+        let orig_depth = depth + tot_del - c - 1 in
+        let subst t = 
+          substitute srk (fun i ->
+            let (q_depth, typ) = Hashtbl.find q_map (orig_depth - i) in
+            mk_var srk (depth - q_depth - 1) typ)
+          t
+        in
+        subst phi
+    | `Not node -> mk_not srk (compute_expr tot_del' depth node)
+    | `Ite (bcond, bif, bthen) ->  
+      mk_ite 
+        srk 
+        (compute_expr tot_del' depth bcond)
+        (compute_expr tot_del' depth bif)
+        (compute_expr tot_del' depth bthen)
+    | `Junct(op, juncts) -> 
+      (mk_junc op) (List.map (compute_expr tot_del' depth) juncts)
+    | `Quantifier (qtyp, name, typ, td, del, node) -> 
+      Hashtbl.add q_map (depth + tot_del' - del - td) (depth, typ);
+      (mk_q qtyp) ~name typ (compute_expr tot_del' (depth + 1) node)
+  in
+  compute_expr 0 0 (Formula.eval srk alg phi)
+
+(* Given a list of equalities, find a candidate
+ * term to substitute in for var 0 *)
+let get_subst_candidate srk eqs = 
+  let unfil_candidates = 
+    List.map (fun (f1, t1, f2, t2) -> 
+        match Term.destruct srk t1, Term.destruct srk t2 with
+        | `Var (ind1, _), `Var (ind2, _) when ind1 = 0 && (not (ind2 = 0)) ->
+          Some t2
+        | `Var (ind1, _), `Var (ind2, _) when (not (ind1 = 0)) && ind2 = 0 ->
+          Some t1
+        | `Var(ind, _), _ when ind = 0 ->
+          if not (DecMap.mem 0 f2)
+          then Some t2
+          else None
+        | _, `Var(ind, _) when ind = 0 -> 
+          if not (DecMap.mem 0 f1) 
+          then Some t1
+          else None
+        | _, _ ->  None) eqs
+  in
+  let filtered_candidates = 
+    List.filter (fun cand ->
+        match cand with
+        | Some _ -> true
+        | None -> false) 
+      unfil_candidates
+  in
+  if List.length filtered_candidates = 0 then None
+  else (List.hd filtered_candidates)
+
+
+(* I found that running this procedure multiple times yields much better
+ * results. *)
+let eq_guided_qe srk phi =
```

Comment:
This is not quantifier elimination -- it doesn't guarantee that quantifiers are eliminated.  It's a simplification routine.